### PR TITLE
drivers: clock_control: nrf: Add option to use external LF source

### PR DESCRIPTION
**Pull request by nordic-krch**
Extended clock configuration to allow usage of external clock
source for nrf52 series.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

**Review by cvinayak — APPROVED** [review]
LGTM

**Comment by ioannisg** [comment]
I am a little bit concerned about the following code in ext/hal/nordic/nrfx/hal/nrf_clock.h:

```
#if defined(NRF52810_XXAA) || \
    defined(NRF5283 2_XXAA) || defined(NRF52832_XXAB) || \
    defined(NRF52840_XXAA)
// Enable support for external LFCLK sources. Read more in the Product Specification.
#define NRF_CLOCK_USE_EXTERNAL_LFCLK_SOURCES
#endif
```
This shows the SOCs for which the new macros, `NRF_CLOCK_LFCLK_Xtal_Low_Swing `and `NRF_CLOCK_LFCLK_Xtal_Full_Swing `are defined. So, the Kconfig dependencies need to reflect this constraint; now there is an inconsistency, because nRF52811 is not in that list, while it is included in the nrf52 series.

**Comment by ioannisg** [comment]
If I understand correctly, if you compile for nRF52811 and set in Kconfig one of the two options, you will get a build error, no?

**Comment by nordic-krch** [comment]
@ioannisg yes, that is correct but then that is a bug in nrfx. I will push nrfx guys to fix it there. Question what should we do about it here. I'm afraid that workaround code here will stay for long even after fix in nrfx. An option is to not use enum but mask directly from bitfields.

**Comment by ioannisg** [comment]
> 
> 
> @ioannisg yes, that is correct but then that is a bug in nrfx. I will push nrfx guys to fix it there. Question what should we do about it here. I'm afraid that workaround code here will stay for long even after fix in nrfx. An option is to not use enum but mask directly from bitfields.

Well, we could also patch the nrfx, here, ourselves; is this a "standard" policy? And of course, wait to be notified about the fix from the nrfx team. This is the simplest solution, but introduces a workaround.

Not use the enum; this implies by-passing the nrfx, and use the MDK directly.

I don't know what is worse; patching or by-passing and "re-inventing" nrfx :) 
After all, nrfx is here to help us....

I vote for a patch in nrfx, but we can quickly discuss it here and conclude. FYI @anangl @carlescufi

**Comment by nordic-krch** [comment]
@ioannisg i've patched nrfx. I thought that it already comes through west as external repo that's why i haven't thought of patching it. Meanwhile I've already prepared patch in nrfx. so it will be in official nrfx in next release.

File: ext/hal/nordic/nrfx/hal/nrf_clock.h [hunk 1]
```diff
@@ -48,9 +48,8 @@ extern "C" {
  * (HFCLK) settings.
  */
 
-#if defined(NRF52810_XXAA) || \
-    defined(NRF52832_XXAA) || defined(NRF52832_XXAB) || \
-    defined(NRF52840_XXAA)
+#if defined(CLOCK_LFCLKSRC_BYPASS_Enabled) && \
+    defined(CLOCK_LFCLKSRC_EXTERNAL_Enabled)
```
Comment (ioannisg) [inline ext/hal/nordic/nrfx/hal/nrf_clock.h]: Alternatively you could just add the NRF52811 in the list. But I am OK if this is effectively identical.

**Comment by ioannisg** [comment]
@nordic-krch thanks

File: ext/hal/nordic/nrfx/hal/nrf_clock.h (hunk 1 above)
Comment (nordic-krch) [reply ext/hal/nordic/nrfx/hal/nrf_clock.h]: imo, it's more future proof

**Review by anangl — APPROVED** [review]
Regarding the changes in nrfx, I'd add a note in the commit message that this is a temporary fix only, and it is supposed to be replaced by the next update of nrfx.

**Comment by ioannisg** [comment]
> 
> 
> Regarding the changes in nrfx, I'd add a note in the commit message that this is a temporary fix only, and it is supposed to be replaced by the next update of nrfx.

Yeah, good point

**Comment by nordic-krch** [comment]
@anangl @ioannisg updated commit message.

**Review by anangl — APPROVED** [review]

**Review by ioannisg — APPROVED** [review]

**Pull request merged**